rfcs/0010-txpool-dos-mitigation.md

```diff
@@ -0,0 +1,238 @@
+        once every lookback window blocks, and lets her consume resources for
+        lookback window blocks, so with a sufficiently small window it's probably
+        impractical, but I'd rather avoid the headache. I don't see a use case for
+        it. Since the account holder is the only one who can spend funds from their
+        account, and since insufficient funds is (almost) the only reason payments
+        can fail, they should never be sending transactions that used to be valid
+        but aren't now.
+    -   Have a minimum transaction fee
+
+        In this design, an attacker can fill the txpool with bogus transactions with
+        fees too low to ever be included in a block for "free". This is bad, and
+        imposing a overall minimum fee that is at least as much as SNARKing a
+        payment costs would prevent it, but figuring out what that minimum should be
+        is complicated, and the attack is only good so long as the pool isn't full,
+        so I think it's not worth it.
+-   What is the impact of not doing this?
+
+    Various vulnerabilities.
+
+## Prior art
+
+Ethereum allows transaction replacement, and multiple pending transactions from
+the same account, without requiring they be valid when run sequentially. If a
+transaction's smart contract errors out, or runs out of gas, miners get to keep
+the fees. So it's sort of a like a deposit. But I don't think they check that
+the there's sufficient balance in the account to cover the sum of transaction
+fees from all pending transactions before accepting another transaction. That
+would be expensive, since the balance of the sending account after a transaction
+runs may be higher than when it started due to smart contracts. There's no
+explicit replacement fee, and there's definitely no deposit system for it. There
+might be a minimum fee increment though. I think they're vulnerable to some of
+these attacks. They evict transactions from the mempool on a lowest-fee-first
+basis.
+
+Bitcoin will do what they call "replace-by-fee" which is the transaction
+replacement thing. They may have a minimum increment, but not a deposit system.
+Transactions are evicted on a lowest-fee-first basis. They allow mempool
+transactions to depend on each other, but not on hypothetical future
+transactions. When a transaction is evicted from the mempool, they also evict
+any transactions that depend on it.
+
+## Unresolved questions
+
+-   What parts of the design do you expect to resolve through the RFC process
+    before this gets merged?
+
+    Decide if it's worth it to do the more complicated thing.
+
+-   What parts of the design do you expect to resolve through the implementation
+    of this feature before merge?
+
+    Nothing comes to mind.
+
+-   What related issues do you consider out of scope for this RFC that could be
+    addressed in the future independently of the solution that comes out of this
+    RFC?
+
+    The SNARK pool has similar concerns.
```

Comment:
I think SNARK pool is a bit different because SNARK works do depend precisely on the current state not just that some account has at least some sort of balance in it. It's harder to nail this down directly, so I think we want a non-zero "block lookback window" for example. However, we should definitely think about the SNARK pool given these concerns you've brought up.